### PR TITLE
Bug 1907309: Add new StorageVersionMigration resourceapply

### PR DESCRIPTION
**Pull request by ihcsim**
This PR adds a new `resourceapply`and `resourceread` for the `migration.k8s.io` 
group. These new functions will be used by the `StaticResourceController` in the
kube-apiserver controller to manage APF resources storage version migration. (
See https://github.com/openshift/cluster-kube-apiserver-operator/pull/1139.)

Test steps:

* Create a 4.7.12 OCP cluster
* Build the `etcdhelper` CLI per instructions [here](https://github.com/openshift/origin/tree/master/tools/etcdhelper)
* Copy the `etcdhelper` binary to the etcdctl container of one of the `etcd` pods:

```sh
oc cp etcdhelper openshift-etcd/etcd-isim-dev-xgfgk-master-0.c.openshift-gce-devel.internal:/opt -c etcdctl
```

* Use `etcdhelper` to confirm the initial `v1alpha1` version of the APF resources:

```sh
$ oc -n openshift-etcd exec etcd-isim-dev-xgfgk-master-0.c.openshift-gce-devel.internal -c etcdctl \
-- /opt/etcdhelper \
-key=/etc/kubernetes/static-pod-certs/secrets/etcd-all-serving/etcd-serving-isim-dev-xgfgk-master-0.c.openshift-gce-devel.internal.key \
-cert=/etc/kubernetes/static-pod-certs/secrets/etcd-all-serving/etcd-serving-isim-dev-xgfgk-master-0.c.openshift-gce-devel.internal.crt \
-cacert=/etc/kubernetes/static-pod-certs/configmaps/etcd-serving-ca/ca-bundle.crt get /kubernetes.io/flowschemas/catch-all --prefix
flowcontrol.apiserver.k8s.io/v1alpha1, Kind=FlowSchema
{
  "kind": "FlowSchema",
  "apiVersion": "flowcontrol.apiserver.k8s.io/v1alpha1",
  "metadata": {
    "name": "catch-all",
    "uid": "494678a7-0ca6-486e-9f47-ebfe5868ee8f",
    "generation": 1,
    "creationTimestamp": "2021-06-02T20:54:54Z",
    # ...
```

* Upgrade the cluster to `4.8.0-fc5`
* Build and deploy custom kube-apiserver operator image to the upgraded cluster
* Confirm that the operator picks up the new manifest:

```sh
$ oc -n openshift-kube-apiserver-operator logs kube-apiserver-operator-64f8bb6d4d-lh6fd | grep -i storageversionmigration
I0603 19:40:49.027872       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-kube-apiserver-operator", Name:"kube-apiserver-operator", UID:"b906bc41-c9e4-4cf8-a06b-b906e5b25086", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'StorageVersionMigrationCreated' Created StorageVersionMigration.migration.k8s.io/flowcontrol-flowschema-storage-version-migration because it was missing
```

* Confirm that the storage version migrator successfully migrated the version:
```sh
$ oc -n openshift-kube-storage-version-migrator logs migrator-68b46b456d-c6psq | grep -i flowschema
I0603 19:40:49.094922       1 kubemigrator.go:110] flowcontrol-flowschema-storage-version-migration: migration running
I0603 19:40:49.536115       1 kubemigrator.go:127] flowcontrol-flowschema-storage-version-migration: migration succeeded
```

* `etcdhelper` should show that the objects are now in the beta version:

```sh
$ oc -n openshift-etcd exec etcd-isim-dev-xgfgk-master-0.c.openshift-gce-devel.internal -c etcdctl \
-- /opt/etcdhelper \
-key=/etc/kubernetes/static-pod-certs/secrets/etcd-all-serving/etcd-serving-isim-dev-xgfgk-master-0.c.openshift-gce-devel.internal.key \
-cert=/etc/kubernetes/static-pod-certs/secrets/etcd-all-serving/etcd-serving-isim-dev-xgfgk-master-0.c.openshift-gce-devel.internal.crt \
-cacert=/etc/kubernetes/static-pod-certs/configmaps/etcd-serving-ca/ca-bundle.crt get /kubernetes.io/flowschemas/catch-all --prefix
flowcontrol.apiserver.k8s.io/v1beta1, Kind=FlowSchema
{                                                                                                                                                 
  "kind": "FlowSchema",                                                                                                                           
  "apiVersion": "flowcontrol.apiserver.k8s.io/v1beta1",                                                                                           
  "metadata": {                                                                                                                                   
    "name": "catch-all",                                                                                                                          
    "uid": "494678a7-0ca6-486e-9f47-ebfe5868ee8f",                                                                                                
    "generation": 1,                                                                                                                              
    "creationTimestamp": "2021-06-02T20:54:54Z",     
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -52,7 +52,7 @@ require (
 	k8s.io/klog/v2 v2.8.0
 	k8s.io/kube-aggregator v0.21.1
 	k8s.io/utils v0.0.0-20201110183641-67b214c5f920
-	sigs.k8s.io/kube-storage-version-migrator v0.0.3
+	sigs.k8s.io/kube-storage-version-migrator v0.0.4
```
Comment (ihcsim) [inline go.mod:55]: bumped to import `sigs.k8s.io/kube-storage-version-migrator/pkg/clients/clientset/fake` in testing.

**Comment by deads2k** [comment]
I expected to see a resourceread too.  Is that not needed for somereason?

**Comment by ihcsim** [comment]
@deads2k `resourceread` added in https://github.com/openshift/library-go/pull/1091/commits/697a3fd2c06cec09a7815e78ee82615bbd70ea11.

File: pkg/operator/staticresourcecontroller/static_resource_controller.go [hunk 2]
```diff
@@ -20,6 +20,8 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
 
+	migrationv1alpha1 "sigs.k8s.io/kube-storage-version-migrator/pkg/apis/migration/v1alpha1"
+
```
Comment (deads2k) [inline pkg/operator/staticresourcecontroller/static_resource_controller.go:24]: nit, no newlines.

Comment (ihcsim) [reply pkg/operator/staticresourcecontroller/static_resource_controller.go]: fixed.

**Comment by deads2k** [comment]
/lgtm
/approve
/hold

holding for proof.

**Comment by deads2k** [comment]
proof appeared to work
/hold cancel

**Pull request merged**